app/controllers/Giraffe.scala

```diff
@@ -124,8 +126,14 @@ class Giraffe(paymentServices: PaymentServices) extends Controller with Redirect
     val maxAmountInLocalCurrency = MaxAmount.forCurrency(countryGroup.currency)
     val creditCardExpiryYears = CreditCardExpiryYears(LocalDate.now.getYear, 10)
 
-    Ok(views.html.giraffe.contribute(pageInfo, maxAmountInLocalCurrency, countryGroup, chosenVariants, cmp, intCmp, creditCardExpiryYears, errorMessage))
-      .withCookies(chosenVariants.variants.map(Test.createCookie):_*)
+    val response = Ok(views.html.giraffe.contribute(pageInfo, maxAmountInLocalCurrency, countryGroup, variant, cmp, intCmp, creditCardExpiryYears, errorMessage))
+
+    val responseWithTests = (for {
+      testId <- request.cookies.get(Test.TestIdCookieName)
+      variantSlug <- request.cookies.get(Test.cookieName(variant))
+    } yield response) getOrElse response.withCookies(Test.testIdCookie(mvtId), Test.variantCookie(variant))
+
+    responseWithTests.discardingCookies(Test.allTests.filterNot(t => t.slug == variant.testSlug) map(t => DiscardingCookie(Test.cookieName(t))): _*)
```

Comment:
I think both are fine. True it would set the cookie, but on a normal workflow you only load that page once anyway :D
No need to change anything though